### PR TITLE
high-level scaffolder

### PR DESCRIPTION
**Pull request by travi**

### CODE REVIEW

**Comment by travi** [comment]
just a very basic start so far, but wanting to make this visible from the beginning

**Comment by travi** [comment]
significantly more useful progress to give an idea of where this is headed.

currently at a decision point of how to pass the decisions to the js-scaffolder. things didnt come together in the way that i initially thought they would, so i'm stepping away to allow some thoughts to stir. i think i have a potential way forward, but dont want to force something non-optimal

**Comment by travi** [comment]
went down the path that i was thinking about and i think it will work well

unfortunately, i'm currently fighting an unrelated issue that i think is `mock-fs` releated but i havent been able to solve yet:
```
TypeError: builtins.forEach is not a function
           at Object.module.exports [as default] (/Users/travi/development/form8ion/eslint-config-extender/node_modules/validate-npm-package-name/index.js:55:12)
           at buildPackageName (/Users/travi/development/form8ion/eslint-config-extender/node_modules/@travi/javascript-scaffolder/lib/index.cjs.js:703:80)
           at scaffold (/Users/travi/development/form8ion/eslint-config-extender/node_modules/@travi/javascript-scaffolder/lib/index.cjs.js:1075:23)
           at async Promise.all (index 1)
           at Object.scaffold$2 [as scaffold] (/Users/travi/development/form8ion/eslint-config-extender/node_modules/@travi/project-scaffolder/lib/index.cjs.js:371:31)
           at World.<anonymous> (/Users/travi/development/form8ion/eslint-config-extender/test/integration/features/step_definitions/common-steps.js:52:5)
```

**Comment by travi** [comment]
to capture some more of the investigation i've gone through:

```
$ npm ls builtins
@form8ion/eslint-config-extender@0.0.0-semantically-released /Users/travi/development/form8ion/eslint-config-extender
├─┬ @travi/javascript-scaffolder@11.9.0
│ └─┬ validate-npm-package-name@3.0.0
│   └── builtins@1.0.3 
└─┬ rollup-plugin-auto-external@2.0.0
  └── builtins@2.0.1 
```

when `console.log`ing `builtins` [from `validate-npm-package-name`](https://github.com/npm/validate-npm-package-name/blob/80313b82a5c7e466556c3a7fb1cb7d7dab92f4be/index.js#L55):

```
{ builtins: [Function] }
```

which suggests that `v2.0.1` is getting pulled in instead of the `v1.0.3` that is actually expected by `validate-npm-package-name`. i'm, so far, unsure if this is behavior introduced by `mock-fs` or `pnpm`, as used by `package-preview`

File: package.json
Comment (trevtrich) [inline package.json:81]: Looks like there's a mix of exact versions and version ranges in dev deps and all version ranges in deps. Might be good to get on same page so I know your strategy to set them if I should add a new one down the road.

File: src/high-level.js
Comment (trevtrich) [inline src/high-level.js:14]: It feels awkward to pass both the option `JavaScript` and the answer `JavaScript` instead of just saying `projectScaffolder(javascriptOptions)`. Almost wonder if there's room for extending the exposed api to have one pathway that doesn't expect the user to make choices but instead just take all input and build. Definitely part of iteration, so not a problem at all, just thought I'd note since I'm the  noob and you've asked me to voice things that feel off as we morph toward some of the new ideas! Definitely like this concept, though 👍

**Review by trevtrich — APPROVED** [review]
I like where it's headed 👍 . One thing that isn'tt clear to me yet at first glance. How does the api to this module get consumed? do i use both the `scaffold` and the `highLevelScaffold`?

File: package.json
Comment (travi) [reply package.json:81]: renovate will enforce that dev deps are exact versions. i may not have set them as such in this branch, but they will get updated. renovate wont enforce either way (with my settings at least) for prod deps, but will honor whichever way it is set. my strategy is a little less than obvious in that context, so happy to chat in a more sync context about this whenever you want

Comment (travi) [reply package.json:81]: high-level answer is that dev deps are for maintainers only, so exact makes sense and forces more frequent updates. prod deps for packages have value in wider ranges to enable deduping on the consumer side. it is important to raise the lower bound when necessary (which a lot of packages dont do and i think cause a lot of the semver frustrations in the community). there are cases where i set prod deps to exact, which mostly come down to forcing updates, but those could be worth chatting more deeply about

File: src/high-level.js
Comment (travi) [reply src/high-level.js:14]: yeah, i can see what you mean, but at least currently, i think that is kind of necessary to keeping the balance between allowing project-scaffolder to be pretty generically usable but enabling orchestration in this type of context. part of the goal of this type of high-level scaffolder is to encapsulate that complexity and abstract it away from consumers tho. certainly open to more discussion here tho.

Comment (travi) [reply src/high-level.js:14]: somewhat related to this, i was initially thinking that i was going to need to provide a constant somewhere that provided a known string for the language key (`JavaScript` in this case) so that it would be predictable enough for this to work. so far all of those are user definable since the config is provided by the consumer (usually a cli at this point). i actually like how this ended up working out because it allows the key to still be consumer definable (the high-level scaffolder in this case) but defined as a constant in this local scope since the necessary usages are all this close to each other

Comment (trevtrich) [reply src/high-level.js:14]: Funny you mention this last point because I had started typing a comment about "might be nice to pin the `JavaScript` string somewhere in the language scaffolder" or something, then realized it was fully encapsulated here within the local scope.

**Comment by travi** [comment]
> How does the api to this module get consumed? do i use both the scaffold and the highLevelScaffold?

yeah, i definitely need to get this updated in the readme example at least. 

the idea is that the `scaffold` export would be added as a package-type plugin option, like [in the `pragmatic-divops/cli`](https://github.com/pragmatic-divops/cli/blob/870e51efe02a5b1c3029638fe61684076e162bc5/src/commands/scaffold/enhanced-scaffolders.js#L22) in the options. then those options would be passed to the high-level scaffolder export (so far i'm running with the name `extendEslintConfig` for the name of the export in this case).

initially, i'm thinking the `extendEslintConfig` would start as a separate cli command that is a peer to `scaffold`, but i think i'd like to see a way for the `scaffold` command to enable choosing a high-level scaffolder as part of its flow somehow

**Comment by trevtrich** [comment]
Ah, okay I see what you're saying. I don't have my head around what you mean by high-level being a peer in the cli command to the core scaffolder, but I think the concept makes sense. I am having a hard time putting together in my head where the responsibilities of one ends and the other starts. I almost wonder if it would be feasible for the higher-level to call itself rather than the project-scaffolder if that makes sense? So: 

```js
import {scaffold} from './scaffold';

export function higherLevel(options) {
  const hardCodedDecisions = {...};
  
  return scaffold(options, hardCodedDecisions);
}
```
I think the fact that `scaffold` writes some files and `highLevel` is responsible for writing other files has my head spinning. Maybe that is just current state as you work toward your final solution though? If so, I can just quit asking questions and wait until you have your thoughts down into the code 😄

**Comment by travi** [comment]
no, feel free to keep the questions coming.

in this case, calling the project scaffolder, and in turn the js-scaffolder, is important because those still handle the heavy lifting of the overall project that is created for the resulting eslint-config. the `scaffold` export from this package is little more than an extension of the js-scaffolder with the details that are very specific to eslint config packages

**Comment by travi** [comment]
i finally got the integration test complexities worked out. didnt end up doing anything that i havent already tried, but this time it worked. hopefully things dont get bumped out of place too easily :\

i think this is ready to merge to the `beta` branch so that we can get some pre-releases publishing. i will make the readme example updates there before merging to `master`

**Comment by trevtrich** [comment]
That is actually a good way to put how i think my head is interpreting the structure here in regard to the high-level vs. the regular `scaffold`. That is: can the `scaffold` method actually do all the lifting? Then the high-level's responsibility would be solely setting args. That way if I came into a project like this for the first time, I could see `scaffold` and use it raw if i wanted to to just extend a random eslint config somewhere on github. But then as a maintainer I can use the high-level one because it already sets a bunch of the stuff I'm going to set anyway (`scope: form8ion`, etc.).

**Comment by travi** [comment]
> can the scaffold method actually do all the lifting?

the intent is that the `scaffold` function does all of the scaffolding lifting _for the specific type of projet intended to be scaffolded by the overall package_. it won't be able to handle the generic project or language scaffolding tasks as those will continue to be handled by the existing generic scaffolders.

the high-level export is intended to _only_ do orchestration to simplify the execution of the project/language scaffolders. i'm sure there will be exceptions to that at some point, but that is the intent at this point.

> I could see scaffold and use it raw if i wanted

yes, for sure, but with some risk of choosing some options within the project/language scaffolders that arent handled well because they are out of scope of the intended behavior of the plugin.

> to just extend a random eslint config somewhere on github

at this point, the `scaffold` plugin is specifically targeted to assume extension of a `form8ion` config

> But then as a maintainer I can use the high-level one because it already sets a bunch of the stuff I'm going to set anyway (scope: form8ion, etc.).

for now the intent isnt to specialize the use, but instead limit the choices to protect from unsupported use cases. for example, with the eslint configs, the choices for unit/integration testing should always be false because there is no value in unit testing these packages using the traditional tools. (there might be value in using [RuleTester](https://eslint.org/docs/developer-guide/working-with-rules#rule-unit-tests) at some point, but ignore that for now). Choosing false for both exposes the question about transpilation, which should also be false. Choosing any of these incorrectly scaffolds the config project in a way that doesnt really make sense for that type of project and can take some effort to undo (and easily miss some details).

the specialization use case is an interesting one that i havent considered too much, but could totally be valid. its simple different than the current goal. truly the "high-level" purpose is orchestration. i'm sure we'll come up with various reasons for orchestration over time. some will be more generally usable and some will be more targetted.

File: package.json
Comment (travi) [reply package.json]: here is the [renovate PR](https://github.com/form8ion/eslint-config-extender/pull/31) to mark the dev-deps as exact, now that we merged to the `beta` branch, which [is watched based on my config](https://github.com/form8ion/renovate-config/blob/53f0682fcb795c3ebc7067dd9e0a07a839d1ba3a/js-package.json#L3)

**Comment by travi** [comment]
> How does the api to this module get consumed? do i use both the scaffold and the highLevelScaffold?

keep an eye on https://github.com/pragmatic-divops/cli/pull/71 for an example. i've added the peer command already, but still need to pull up the common options and decisions

**Pull request merged**
